contracts/cw-named-groups/src/contract.rs

```diff
@@ -193,36 +193,26 @@ pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
 }
 
 fn query_dump(deps: Deps) -> StdResult<DumpResponse> {
-    let mut groups_map: HashMap<String, Vec<String>> = HashMap::new();
-
     // Map groups to contained addresses.
-    GROUPS
+    let groups = GROUPS
         .groups_to_addresses
         .keys(deps.storage, None, None, Order::Ascending)
-        .try_for_each::<_, StdResult<()>>(|element| {
-            let element = element?;
-            let group_name = element.0;
-            let address = element.1;
-            let mut addresses = Vec::new();
-            if groups_map.contains_key(&group_name) {
-                addresses = groups_map.get(&group_name).unwrap().to_vec();
-            }
-            addresses.push(address.to_string());
-            groups_map.insert(group_name.to_string(), addresses);
-            Ok(())
-        })?;
-
-    // Convert groups map to dump response.
-    let mut dump: Vec<Group> = Vec::new();
-    groups_map.into_iter().for_each(|element| {
-        let group: Group = Group {
-            name: element.0,
-            addresses: element.1,
-        };
-        dump.push(group);
-    });
-
-    Ok(DumpResponse { groups: dump })
+        .collect::<StdResult<Vec<(String, Addr)>>>()?;
+
+    let (latest_name, _) = groups.first().ok_or(StdError::generic_err("not found"))?;
+
+    let mut dump: Vec<Group> = vec![];
+    let mut tmp: Vec<String> = vec![];
+    for (name, addr) in groups {
+        if name == *latest_name {
+            tmp.push(addr.to_string());
+        }
+        else {
+            dump.push(Group{ name, addresses: tmp })
```

Comment:
Given a set of groups of `{('water', 'fish'), ('water', 'plankton'), ('land', 'dog')}` it's not clear to me this works.

1. Set `latest_name = 'water'`.
2. While `latest_name == name` push to tmp => tmp becomes `['plankton', 'dog']`.
3. `latest_name != 'land'` so we add `Group { name: 'land', addresses: ['plankton', 'dog'] }`

Am I misunderstanding this logic? Haven't looked much at this contract so could be missing something.

Reply:
Think likely what we want to do here is
```rust
dump.push(Group{ name: latest_name, addresses: tmp });
latest_name = name;
tmp = vec![addr.to_string()]
```

Reply:
I think ekez's proposed logic is correct. 

If the Map structure supported some kind of GROUP BY operation for composite keys where we could collapse entries by a prefix, that would be cool, but perhaps too niche of an operation. 

But also I'm curious, why is it so important to avoid using a hash map here since we're not persisting it to storage? How much of an optimization gain would it be?

Reply:
HashMap builds runs the keys through a hashing algorithm. Hashing is somewhat an expensive operation for this desired logic and can be easily done by pushing data to vectors.